src/engagement_db_coda_sync/lib.py

```diff
@@ -1,15 +1,106 @@
 from core_data_modules.data_models import Message as CodaMessage, Label, Origin
 from core_data_modules.logging import Logger
 from core_data_modules.util import TimeUtils
 from engagement_database.data_models import HistoryEntryOrigin
+from storage.google_cloud import google_cloud_utils
 
 from src.engagement_db_coda_sync.sync_stats import CodaSyncEvents, EngagementDBToCodaSyncStats
 
 log = Logger(__name__)
 
 
+def get_coda_users_from_gcloud(coda_dataset_users_file_url, google_cloud_credentials_file_path):
+    return json.loads(google_cloud_utils.download_blob_to_string(
+        google_cloud_credentials_file_path, coda_dataset_users_file_url
+    ))
+
+
+def ensure_coda_datasets_up_to_date(coda, coda_config, google_cloud_credentials_file_path): 
+    """
+    Ensures coda datasets are up to date based on coda configuration. 
+
+    :param coda: Coda instance to add the message to.
+    :type coda: coda_v2_python_client.firebase_client_wrapper.CodaV2Client
+    :param coda_config: Configuration for the update.
+    :type coda_config: src.engagement_db_coda_sync.configuration.CodaSyncConfiguration
+    :param google_cloud_credentials_file_path: Path to a Google Cloud service account credentials file 
+                                               to use to access the credentials bucket.
+    :type google_cloud_credentials_file_path: str
+    """
+    all_datasets_have_user_file_url = all(
+        dataset_config.coda_dataset_users_file_url is not None for dataset_config in coda_config.dataset_configurations)
+
+    default_user_ids = []
+    if not all_datasets_have_user_file_url:
+        assert coda_config.project_users_file_url is not None, \
+         f"Specify user_ids for coda datasets in CodaDatasetConfiguration or user_ids for this project in CodaSyncConfiguration"
+        default_user_ids = get_coda_users_from_gcloud(coda_config.project_users_file_url, google_cloud_credentials_file_path)
+
+    ws_correct_dataset_code_scheme = coda_config.ws_correct_dataset_code_scheme
+    for dataset_config in coda_config.dataset_configurations:
+        user_ids = []
+        if dataset_config.coda_dataset_users_file_url:
+            user_ids.append(get_coda_users_from_gcloud(dataset_config.coda_dataset_users_file_url))
+        else:
+            user_ids = default_user_ids
+
+        coda_user_ids = coda.get_dataset_user_ids(dataset_config.coda_dataset_id)
+        if coda_user_ids:
+            user_ids = list(set(user_ids) - set(coda_user_ids))
+            coda.update_dataset_user_ids(dataset_config.coda_dataset_id, user_ids)
+        else:
+            coda.set_dataset_user_ids(dataset_config.coda_dataset_id, user_ids)
+
+        repo_code_schemes = []
+        for code_scheme_config in dataset_config.code_scheme_configurations:
+            assert code_scheme_config.coda_code_schemes_count < 5, \
+                "`code_scheme_configuration.coda_code_schemes_count` exceeds number of columns that can be displayed in coda"
```

Comment:
I'm not convinced we really need this check - but if we do, it's more complicated than this, as you need to compute the total number of code schemes that would be shown in Coda, not just look at each in turn like this.

Reply:
I've removed this for now as I think of the best way to do this.